cpp/src/arrow/compute/function.h

```diff
@@ -159,6 +159,14 @@ struct ARROW_EXPORT FunctionDoc {
   static const FunctionDoc& Empty();
 };
 
+/// \brief An executor of a function with a preconfigured kernel
+struct ARROW_EXPORT FunctionExecutor {
+  virtual ~FunctionExecutor() {}
+  /// \brief Execute the preconfigured kernel with arguments that must fit it
+  virtual Result<Datum> Execute(const std::vector<Datum>& args,
+                                int64_t passed_length = -1) = 0;
```

Comment:
I would expect the `ExecContext` to be passed here, rather than on kernel resolution.

Reply:
As code now, both the exec context and the function options are used in kernel initialization, in `FunctionExecutorImpl.Init`. If both are moved to kernel execution, then kernel initialization would be invoked on each execution, which IIRC is not good for my UDF purposes. I guess we could separate kernel resolution from kernel initialization, and allow a choice if which one would be reused across executions.

Reply:
> I guess we could separate kernel resolution from kernel initialization

We should.